### PR TITLE
Refactor internal scope handling by introducing Scope enum

### PR DESCRIPTION
**Pull request by nicoddemus**
This replaces the internal use of string literals and "scopenum"
to a proper Scope enum, which also centralizes indexes, getting
higher scopes, get next scope in the order, etc. Another benefit
is that it helps the type checker by introducing a proper type,
and improve readability (IMHO).

This is _mostly_ an internal change, however due to historical
reasons, the API of the following _internal_ objects has
changed slightly:

* `CallSpec2`: `_arg2scopenum` renamed to `_arg2scope` for consistency.
* `FixtureRequest`:
  Previously contained a `scope: str` attribute.
  Changed attribute to `_scope: Scope`, with a backward compatible read-only property `scope -> str`.
  This one might cause some confusion, because the internal attribute has a different type
  than the public property of the same name.
* `SubRequest.__init__` parameter changed from `_Scope` to `Scope`.

### CODE REVIEW

File: src/_pytest/scope.py
Comment (bluetech) [inline src/_pytest/scope.py:4]:
```suggestion
Those are defined here, instead of in the 'fixtures' module, because
```

File: src/_pytest/unittest.py
Comment (bluetech) [inline src/_pytest/unittest.py:144]: Perhaps this can be done in Scope itself by overriding `value` (or just adding `value: _ScopeName` to the class)? (Didn't check if that's possible)

File: src/_pytest/fixtures.py
Comment (bluetech) [inline src/_pytest/fixtures.py:454]: As long as we're not exposing `Scope` and `_scope` is private, I don't think we should document it as backward-compat, since we don't expect anyone to use something else.

File: src/_pytest/fixtures.py
Comment (bluetech) [inline src/_pytest/fixtures.py:1016]: Same here re. backward-compat designation.

**Review by bluetech — COMMENTED** [review]
Very nice cleanup. I'll need to look again when I get a bit more time but LGTM from just reading once.

One thing is the `index` - can we get rid of it? If I'm reading correctly, it's only used for sorting/comparing `Scope`s, and for the indentation bit in `setuponly`. Can we instead implement ordering for `Scope` and compare the values directly? For `setuponly` can do something ad-hoc.

File: src/_pytest/fixtures.py
Comment (nicoddemus) [reply src/_pytest/fixtures.py:454]: Agreed. But on second thought perhaps we can ditch the property completely? I find it a bit confusing to have `scope: str` and `_scope: Scope`. If we don't expect to support backward compatibility here, given it is an internal object, might make sense to just drop the property entirely.

Comment (bluetech) [reply src/_pytest/fixtures.py:454]: I don't think it's an internal object. I'll be surprised if no one is using `request.scope`. It is also [documented](https://docs.pytest.org/en/latest/reference/reference.html#pytest.FixtureRequest.scope).

One option is to use an `Enum` which also derives from `str`, then it will be compatible. From my experience, these cause a lot of confusion, but may make sense here.

Comment (nicoddemus) [reply src/_pytest/fixtures.py:454]: Hmm then I think the backward compatibility bit on the docstring makes sense TBH, we have that difference only because it is part of the public API.

Comment (bluetech) [reply src/_pytest/fixtures.py:454]: What I mean to say is that we shouldn't designate something as backward compatibility when it doesn't have a replacement. The new attribute is not a replacement because it is private.

Comment (nicoddemus) [reply src/_pytest/fixtures.py:454]: Hmm not sure I follow, but I don't feel strongly about it either. Can you tell what docstring exactly you would like to see there then? I'm happy to change it.

File: src/_pytest/scope.py
Comment (RonnyPfannschmidt) [inline src/_pytest/scope.py:21]: i just looked up the details of enum

lets use the example of https://docs.python.org/3/library/enum.html#when-to-use-new-vs-init
and store the name and the index togethe so its
```
 Session = ("session", 1)
 Package = ("package", 2)
```

going overboard we might want to use `auto()` + its heleprs to get al lthe things in

for future reference we might want to add the definition scope one before function (its intended to eventually come in between the collectors and the functions to contain parameterization)

Comment (nicoddemus) [reply src/_pytest/scope.py:21]: Why do you feel having the index there is important? See the comment from @bluetech:

https://github.com/pytest-dev/pytest/pull/8913#pullrequestreview-708086309

I was going to try to get rid of that index altogether. Do you feel it is necessary to have it there for other reasons than sorting them?

File: src/_pytest/scope.py
Comment (RonnyPfannschmidt) [inline src/_pytest/scope.py:73]: part of me wonders if the declaration of next_scope shouldnt just be a plain old mapping

we could pretty much do with a ordered mapping there

File: src/_pytest/fixtures.py
Comment (bluetech) [reply src/_pytest/fixtures.py:454]: > Hmm not sure I follow

It's probably me. Maybe your intention is to *remove* `request.scope` without replacement? If so, I'm not sure about it, I think it's useful.

> Can you tell what docstring exactly you would like to see there then

Can use the existing docstring

> Scope string, one of "function", "class", "module", "session".

although it's missing "package" so should add that.

File: src/_pytest/scope.py
Comment (nicoddemus) [reply src/_pytest/scope.py:73]: If you prefer something more explicit, probably we can then use this:

```python
@lru_cache(maxsize=None)
def _next_scope(scope: Scope) -> Scope:
    if scope is Scope.Function:
        raise ValueError("Function is the bottom scope")
    if scope is Scope.Session:
        return Scope.Package
    if scope is Scope.Package:
        return Scope.Module
    if scope is Scope.Module:
        return Scope.Class
    if scope is Scope.Class:
        return Scope.Function
    assert_never(scope)
```

So `assert_never` will fail if we ever add a new scope and forget to update this method.

However TBH I think the current implementation is OK, but don't have strong feelings either.

Comment (RonnyPfannschmidt) [reply src/_pytest/scope.py:73]: no, what i meant is 

```py
NEXT_SCOPES = {
  Scope.Session: Scope.Package,
  ....
  }
```

Comment (nicoddemus) [reply src/_pytest/scope.py:73]: I know, but I was proposing the above because then we can use `assert_never`.

However we still need a function to be able to raise a proper error when `Scope.Function` is given, that's why I'm not sure having a mapping buys us much:

```
@lru_cache(maxsize=None)
def _next_scope(scope: Scope) -> Scope:
    if scope is Scope.Function:
        raise ValueError("Function is the bottom scope")
    return _NEXT_SCOPES[scope]
```

Comment (RonnyPfannschmidt) [reply src/_pytest/scope.py:73]: we can also drop the lru cache

drop the data spec into a dict, leave off the lru cache, and actually, if we use _next_scopes[scope], then we get a lookuperror by default as well (if we use it internally, then we dont need the extra text in the value error)

File: src/_pytest/scope.py
Comment (RonnyPfannschmidt) [inline src/_pytest/scope.py:35]: bascially this one is the public api can can raise a better value error

File: src/_pytest/scope.py
Comment (RonnyPfannschmidt) [inline src/_pytest/scope.py:62]: if we declare NEXT_SCOPE as mapping, then its keys will be equal to the HIGH_SCOPES list

i believe all modern pythons also have the order fixed

File: src/_pytest/scope.py
Comment (RonnyPfannschmidt) [inline src/_pytest/scope.py:66]: similarly  this can be declared as `SCOPE_TO_INDEX = {s: i for s, i in enumerate(Scope)}` with the the caveat that i always forget the tuple order of enumerate

File: src/_pytest/scope.py
Comment (nicoddemus) [reply src/_pytest/scope.py:73]: Do you think using a mapping is superior? One thing I don't like is that I have to duplicate the names of the scopes. The current implementation doesn't require that.

Can you clarify why do you prefer the mapping?

Comment (RonnyPfannschmidt) [reply src/_pytest/scope.py:73]: i would use the mapping as its a superior lookup stucture

i would actually declare scopes/next_scopes in terms of dictionary expressions

aka ```py
SCOPE_INDEXES ={s:i for ....}
NEXT_SCOPES = {s: s2 for s, s2 in zip(...), ...}
```

File: src/_pytest/unittest.py
Comment (nicoddemus) [reply src/_pytest/unittest.py:144]: Hmm good idea, but unfortunately it does not work. 😞

File: src/_pytest/scope.py
Comment (nicoddemus) [reply src/_pytest/scope.py:35]: You were right in the other comment, this is used internally only. 👍

**Comment by nicoddemus** [comment]
> One thing is the index - can we get rid of it?

I did this locally, but I found the results surprising/confusing: if we "order" scopes, we would expect this to be true:

```python
assert Scope.Session > Scope.Package
```

However that's the inverse of what we have today with "scope num", because `Session(index:0) < Package(index:1)`. Of course we can change the code which uses this to accommodate that, but I would prefer for this PR to just "improve readability" rather than change the underlying "scope num" concept.

**Comment by nicoddemus** [comment]
Implemented most of the suggestions, please see comments. 👍

**Review by bluetech — COMMENTED** [review]
Great work @nicoddemus. Here are my comments:

- Making `scope.value` have type `_ScopeName` can be done by annotating each of the enum entires with `_ScopeName`, e.g. `Function: _ScopeName = "function"`.

- I don't like the name "next" very much because it is can either mean next lower or next higher.
  Therefore I suggest we replace `next()` with `next_lower()` and `next_higher()` (even if one of them is not used, just for consistency).

- Regarding the `index`/ordering, I still think this is a good opportunity to get rid of this redundant concept.
  E.g. consider the `_check_scope` (negative) condition `if requested_scope.index() > invoking_scope.index(): fail`.
  To me it reads backwards. My preference would be:
  - Change the order of items in the enum to `Function -> .. -> Session` rather than the other way around (as it is now).
    If we define an order, I think this is the more intuitive one.
  - Implement the ordering and document it, e.g.

        ```
        Scopes are ordered from lower to higher, that is

                      ->>> higher ->>>
        Function < Class < Module < Package < Session
                      <<<- lower  <<<-
        ```
   - Drop the index concept, use direct comparisons instead.

  (I can do this in a follow up PR, if you agree with the idea)

- Let's not rely on `%r` of the enum in `_check_scope` because recent experience shows it is not stable.

- Should add a changelog entry with the last paragraph of the PR description -- looking at the plugin corpus I see several plugins which use e.g. `_arg2scopenum` and will need to adapt.
  Just to be clear, since these are all internal changes, I'm only suggesting a courtesy changelog to let the plugin authors know about the change, nothing more.

- I think we should take the opportunity to drop the underscore from `_ScopeName` and make it public. It is a stable type and is already opaquely referred to in the API reference.
  Since `Literal` is only available since Python 3.8, we will need some version and `TYPE_CHECKING` to turn it to just `str` on <3.8 runtime, but shouldn't be a problem.
  (I can do this in a follow up PR, if you agree with the idea)

**Comment by nicoddemus** [comment]
Thanks @bluetech for the great review!

Implemented almost everything except making `_ScopeName` public, could you do that in a separate PR as you offered? Thanks!

@RonnyPfannschmidt in order to implement ordering and `next_higher` and `next_lower`, I went with using the Enum ordering rather than the dict approach that was used before, felt more straightforward and avoids repetition, but those are internal details and we have tests, so feel free to open a separate PR if you want the implementation changed.

File: src/_pytest/scope.py
Comment (bluetech) [inline src/_pytest/scope.py:56]: We should apply `@functools.total_ordering` to the class so that all of the comparisons are implemented.

**Review by bluetech — APPROVED** [review]
One small comment, with it, LGTM!

File: src/_pytest/scope.py
Comment (RonnyPfannschmidt) [inline src/_pytest/scope.py:40]: how often do we use this in lookup? i see us using those all over the time regenerating them all the time

off hand i cant tell of we have this in inner loops, but i'd really like to keep it out of loops

in particular list(Scopes) looks like something we should just make a constant
same goes for the index lookups

File: src/_pytest/python.py
Comment (RonnyPfannschmidt) [inline src/_pytest/python.py:1282]: i propose using `min((the_generator), default=Scope,Function)`

File: src/_pytest/scope.py
Comment (nicoddemus) [reply src/_pytest/scope.py:40]: Very rarely I think, but initially I had those implemented as external utility functions and cached. I'm happy to implement it like that again, if you agree.

File: src/_pytest/python.py
Comment (nicoddemus) [reply src/_pytest/python.py:1282]: Sure, will do.

File: src/_pytest/scope.py
Comment (nicoddemus) [reply src/_pytest/scope.py:56]: Oops completely missed that, thanks!

File: src/_pytest/scope.py
Comment (nicoddemus) [reply src/_pytest/scope.py:44]: Done slightly different than the other time, let me know what you think.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus i wonder if perhaps a integer enum should be used

Then the name would be the name and the value would be a comparable number

As far as I understood those are automatically sortable

Unfortunately I won't be able to try this hypothesis today

**Comment by nicoddemus** [comment]
Hmm it is an interesting possibility, however if I understand correctly then `Enum.value` would return an integer, correct?

```python
>>> from enum import IntEnum
>>> class Scope(IntEnum):
...   Function = 0
...   Class = 1
...
>>> Scope.Function
<Scope.Function: 0>
>>> Scope.Function.value
0
>>> Scope.Function.name
'Function'
```

Then we will need to somehow map the values to their known string names (`"function"`, `"class"`, etc).

However I'm not sure that's the right call: we will be trading the current abstraction of having enumerated scopes with known string values, but subclassing from an `IntEnum` assigns an index to each scope as a prominent feature, while I feel their ordering/index is more of an internal detail.

**Comment by bluetech** [comment]
I agree, better to keep it a plain Enum and not inherit from int or str.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus as the relation between the known name, and the enum name is `lower/capitalize` i don`t see a strong problem there

we could also just choose to  use the legacy names instead of capitalized names 

additionally we could choose to accept the enum in all places that accept the name

**Comment by nicoddemus** [comment]
Not sure. To me, the "scope number" aspect of scopes is an implementation detail, and we shouldn't really make it a part of the API. The moment we declare that `Scope.Function` is the same as `1` (because we make that declaration by subclassing `IntEnum`), we are then casting in stone that, which I'm not sure it is beneficial.

Just to be clear, why do you believe the change you are proposing would be beneficial? Just to reduce any code duplication, or are you envisioning something more?

**Comment by RonnyPfannschmidt** [comment]
i think i have to test out some of the details to be sure, my idea should be postponed for now

also integer valued enum and integer enum are not quite the same (i would use a normal enum but have integers as values)

**Pull request merged**
